inc/saiipsec.h

```diff
@@ -0,0 +1,1047 @@
+     */
+    SAI_IPSEC_PORT_STAT_TX_NON_IPSEC_PKTS,
+
+    /**
+     * @brief Packets dropped after receive MAC and before IPsec SA processing.
+     * This could be due to malformed header, buffer overrun, etc
+     */
+    SAI_IPSEC_PORT_STAT_RX_ERROR_PKTS,
+
+    /**
+     * @brief Packets mapped to an SA for IPsec processing.
+     */
+    SAI_IPSEC_PORT_STAT_RX_IPSEC_PKTS,
+
+    /**
+     * @brief Non-IPsec packets that pass through this port.
+     */
+    SAI_IPSEC_PORT_STAT_RX_NON_IPSEC_PKTS,
+} sai_ipsec_port_stat_t;
+
+/**
+ * @brief Attribute Id for sai_ipsec_sa
+ */
+typedef enum _sai_ipsec_sa_attr_t
+{
+    /**
+     * @brief Start of IPsec Security Association attributes
+     */
+    SAI_IPSEC_SA_ATTR_START,
+
+    /**
+     * @brief IPsec direction
+     *
+     * @type sai_ipsec_direction_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_IPSEC_SA_ATTR_IPSEC_DIRECTION = SAI_IPSEC_SA_ATTR_START,
+
+    /**
+     * @brief IPsec object id
+     *
+     * @type sai_object_id_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @objects SAI_OBJECT_TYPE_IPSEC
+     */
+    SAI_IPSEC_SA_ATTR_IPSEC_ID,
+
+    /**
+     * @brief SA status.
+     *
+     * @type sai_ipsec_sa_status_t
+     * @flags READ_ONLY
+     */
+    SAI_IPSEC_SA_ATTR_STATUS,
+
+    /**
+     * @brief Externally assigned SA Index value for this Security Association.
+     * Used only when SAI_IPSEC_ATTR_EXTERNAL_SA_INDEX_ENABLE == true.
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 0
+     */
+    SAI_IPSEC_SA_ATTR_EXTERNAL_SA_INDEX,
+
+    /**
+     * @brief SA Index value for this Security Association.
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_IPSEC_SA_ATTR_SA_INDEX,
+
+    /**
+     * @brief List of IPsec ports for this SA.
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_IPSEC_PORT
+     * @default empty
+     */
+    SAI_IPSEC_SA_ATTR_IPSEC_PORT_LIST,
+
+    /**
+     * @brief SPI value for this Security Association, carried in ESP header.
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_IPSEC_SA_ATTR_IPSEC_SPI,
+
+    /**
+     * @brief Enable 64-bit ESN (vs 32-bit SN) for this Security Association
+     *
+     * @type bool
+     * @flags CREATE_ONLY
+     * @default true
+     */
+    SAI_IPSEC_SA_ATTR_IPSEC_ESN_ENABLE,
+
+    /**
+     * @brief Cipher suite for this SA.
+     *
+     * @type sai_ipsec_cipher_t
+     * @flags CREATE_ONLY
+     * @default SAI_IPSEC_CIPHER_AES256_GCM16
+     */
+    SAI_IPSEC_SA_ATTR_IPSEC_CIPHER,
+
+    /**
+     * @brief IPsec Traffic Encryption Key used for encryption/decryption.
+     * Network Byte order. AES128 uses only Bytes 16..31.
+     *
+     * @type sai_macsec_sak_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     */
+    SAI_IPSEC_SA_ATTR_MACSEC_SAK,
```

Comment:
Why is MACSEC used here in the attribute name?

Reply:
This is a cipher-suite object independent of MACSEC and IPSEC.  The object was incorrectly named in MACSEC_SAK in saimacsec.  Now the choices are:
1. Fix the naming in saimacsec and share the same object definition.  But this causes saimacsec code churn.
2. Fix the name in IPsec only and have duplicate definition of identical objects
I would like broader input in this forum before settling this.

Reply:
MACSEC and IPSEC use an identical object dictated by AES definition.  Unfortunately the object was named MACSEC_SAK in saimacsec.  Now the choices are:
1. Fix saimacsec to use a common object definition.  But this causes macsec code churn.
2. Live with this strange name (SAI_IPSEC_SA_ATTR_MACSEC_SAK) in saiipsec.
3. Fix the name in saiipsec, but live with the poor naming in saimacsec
I would like to get broader input for this.

Reply:
As per agreement in meeting, went with #3(separate but identical object definition for saiipsec with its own serialization/deserialization functions).